### PR TITLE
Remove hparams from checkpoint tests

### PR DESCRIPTION
**Pull request by hanlint**
This PR refactors the checkpoint tests to not use hparams objects, for the following tests:
* test_autoresume
* test_load_weights
* test_load_weights_object_store
* test_save_overwrite
* test_checkpoint
* test_different_run_name

These tests are pretty critical to our checkpointing infra, so I ask the reviewers to please review the logic carefully and make sure the tests are functionally identical to before the refactor.

The existing tests had some functionality (e.g. testing event calls) that were more appropriate to be put in `test_events.py`

### CODE REVIEW

File: tests/trainer/test_checkpoint.py [hunk 1]
```diff
@@ -230,275 +182,48 @@ def test_ignore_params(remove_field_paths: List[List[str]], filter_params: List[
...
-    # load only model weights
-    second_trainer_hparams.load_path = checkpoint_a_file_path[0]
-    second_trainer_hparams.load_weights_only = True
-    second_trainer_hparams.load_strict_model_weights = True
-    # setup a new optimizer
-    second_trainer_hparams.optimizers = AdamWHparams()
-
-    # setup a new LR scheduler
-    assert isinstance(second_trainer_hparams.max_duration, str)
-    second_trainer_hparams.schedulers = [CosineAnnealingScheduler(t_max=second_trainer_hparams.max_duration)]
```
Comment (hanlint) [inline tests/trainer/test_checkpoint.py:269]: I removed the check for the different scheduler type, as I wasn't sure what its purpose is in this test.. does keeping the same scheduler somehow change how the weights are loaded?

File: tests/trainer/test_checkpoint.py [hunk 2]
```diff
@@ -230,275 +182,48 @@ def test_ignore_params(remove_field_paths: List[List[str]], filter_params: List[
...
-
-    # Create checkpoint in seperate folder to load. Optionally delete original checkpoint by moving it.
-    if delete_local_checkpoint:
-        shutil.move(checkpoint_a_folder, checkpoint_b_folder)
-    else:
-        shutil.copytree(checkpoint_a_folder, checkpoint_b_folder, symlinks=True)
-    # Recreate symlink in new folder
-    new_latest_path = os.path.join(checkpoint_b_folder, latest_checkpoint)
-    os.remove(new_latest_path)
-    os.symlink(final_checkpoint, new_latest_path)
```
Comment (hanlint) [inline tests/trainer/test_checkpoint.py:350]: I also removed this logic here, and removed the use of a `second` folder. Am I missing something here?

**Comment by hanlint** [comment]
I bit unsure why these tests are failing.. unable to reproduce locally and seem unrelated to my PR's changes.

File: tests/test_events.py [hunk 3]
```diff
@@ -2,10 +2,130 @@
...
+from tests.common import RandomClassificationDataset, SimpleModel
+from tests.common.events import EventCounterCallback
 
 
 @pytest.mark.parametrize('event', list(Event))
 def test_event_values(event: Event):
     assert event.name.lower() == event.value
+
+
+@pytest.mark.skip
```
Comment (mvpatel2000) [inline tests/test_events.py:20]: Why is there a skip?

File: tests/test_events.py [hunk 4]
```diff
@@ -2,10 +2,130 @@
...
+class TestEventCalls:
```
Comment (mvpatel2000) [inline tests/test_events.py:21]: Why does this need to be a class?

File: tests/trainer/test_checkpoint.py [hunk 5]
```diff
@@ -230,275 +135,49 @@ def test_ignore_params(remove_field_paths: List[List[str]], filter_params: List[
...
-    del device  # unused. Set automatically
-    if not isinstance(composer_trainer_hparams.train_dataset, SyntheticHparamsMixin):
-        pytest.skip('Checkpointing tests require synthetic data')
-        return
-    if not isinstance(composer_trainer_hparams.val_dataset, SyntheticHparamsMixin):
-        pytest.skip('Checkpointing tests require synthetic data')
-        return
-    if use_object_store:
-        pytest.importorskip('libcloud')
+class TestCheckpointLoading:
```
Comment (mvpatel2000) [inline tests/trainer/test_checkpoint.py:138]: same question, why does this need to be a class?

File: tests/trainer/test_checkpoint.py (hunk 1 above)
Comment (mvpatel2000) [reply tests/trainer/test_checkpoint.py:269]: I believe it's to test `load_weights` doesn't override other params, eg scheduler? I'm not super sure. it might be nice to include in the test though to verify `load_weights` doesn't touch other parts

File: tests/trainer/test_checkpoint.py (hunk 2 above)
Comment (mvpatel2000) [reply tests/trainer/test_checkpoint.py:350]: So what's happening is `autoresume` first checks for local checkpoints, and if there aren't any, checks the object stores for checkpoints to resume from. 

The logic here is to say: 1) backup checkpoint to a second folder so the original trainer can reload it and 2) (optionally) delete the checkpoint so we can test if creating a new trainer reloads from object store if local checkpoint is missing.

We don't need the second folder anymore since we can keep an instantiated copy of the original trainer to use for comparisons. However, we do still need to check if deleting the original file (`delete_local_checkpoint`) works. 

I believe your refactor looks like it handles this correctly to me :)

File: tests/trainer/test_checkpoint.py [hunk 6]
```diff
@@ -507,290 +186,333 @@ def test_checkpoint_with_object_store_logger(
...
-
-    assert_checkpoints_equivalent(
-        checkpoint_file_a=first_trainer_final_checkpoint_filepath,
-        checkpoint_file_b=second_trainer_final_checkpoint_filepath,
+        # TODO (mihir): should be checking the entire state
+        # not just the model?
+        self._assert_weights_equivalent(
+            trainer_1.state.model,
+            trainer_2.state.model,
+        )
```
Comment (mvpatel2000) [inline tests/trainer/test_checkpoint.py:280]:
```suggestion
        self._assert_weights_equivalent(
            trainer_1.state.model,
            trainer_2.state.model,
        )
```
LGTM! yes, it should be comparing the entire state, since with `autoresume` there should be no difference from continuing the original vs. restarting from the new one

File: tests/trainer/test_checkpoint.py [hunk 7]
```diff
@@ -507,290 +186,333 @@ def test_checkpoint_with_object_store_logger(
...
+        trainer_1.fit()
+        trainer_1.close()
+
+        trainer_2 = self.get_trainer(
+            load_path=os.path.join('first', 'ep2.pt'),
+            seed=67890,
+        )
+
+        assert trainer_1.state.run_name != trainer_2.state.run_name
```
Comment (mvpatel2000) [inline tests/trainer/test_checkpoint.py:298]:
```suggestion
    def test_different_run_names(self):

        trainer_1 = self.get_trainer(
            save_folder='first/',
            seed=12345,
        )
        trainer_1.fit()
        trainer_1.close()

        trainer_2 = self.get_trainer(
            load_path=os.path.join('first', 'ep2.pt'),
            seed=12345,
        )

        assert trainer_1.state.run_name != trainer_2.state.run_name
```
I believe it should now generate different names even if the seed is the same. It might be worth changing this test to use the same seed and then deleting the new test file I added `tests/trainer/test_run_name.py` to simplify tihs

File: tests/trainer/test_checkpoint.py [hunk 8]
```diff
@@ -507,290 +186,333 @@ def test_checkpoint_with_object_store_logger(
...
+        )
+        trainer_1.fit()
+        trainer_1.close()
+
+        if save_overwrite:
+            ctx = contextlib.nullcontext
+        else:
+            ctx = functools.partial(pytest.raises, FileExistsError)
+
+        with ctx():  # expect FileExistsError if save_overwrite=False
```
Comment (mvpatel2000) [inline tests/trainer/test_checkpoint.py:316]: (minor) nit: would prefer to one-line or define `ctx = None` at top. Surprised lint isn't complaining about this.

**Review by mvpatel2000 — COMMENTED** [review]
Wow!! Amazing work 🚀 . This is _so_ much cleaner without Hparams. I left a few comments, biggest one being I don't get the reason why we have classes used everywhere.

Because this is very subtle and critical code, I would prefer to do a second pass once comments are resolved, especially because I might spot something on a second read through I missed the first time.

File: tests/trainer/test_checkpoint.py (hunk 8 above)
Comment (hanlint) [reply tests/trainer/test_checkpoint.py:316]: Since it's an `if .., else`, the `ctx = None` would always be overridden, so it should be ok?

Comment (mvpatel2000) [reply tests/trainer/test_checkpoint.py:316]: My personal preference is to always put a `ctx=None` even if its always overridden to make scope clear. Even though python doesn't scope indents, most people tend to think in those scope levels, so its nice to be explicit about it. With that said, this is a minor nit you are free to ignore without further discussion :)

File: tests/trainer/test_checkpoint.py (hunk 1 above)
Comment (hanlint) [reply tests/trainer/test_checkpoint.py:269]: Good point, I added a new test to check that the statefulcallback was not touched.

File: tests/trainer/test_checkpoint.py (hunk 8 above)
Comment (hanlint) [reply tests/trainer/test_checkpoint.py:316]: OK understood, added the `ctx = None`.

File: tests/test_events.py [hunk 9]
```diff
@@ -56,9 +55,6 @@ def get_trainer(self, **kwargs):
     @pytest.mark.parametrize('device,deepspeed_zero_stage', [
         pytest.param('cpu', None, id='cpu-ddp'),
         pytest.param('gpu', None, id='gpu-ddp', marks=pytest.mark.gpu),
-        pytest.param('gpu', 0, id='deepspeed-zero0', marks=pytest.mark.gpu),
```
Comment (hanlint) [inline tests/test_events.py]: There is an issue with deepspeed here -- the tests pass, but somehow the errors overflow into other tests.

**Comment by hanlint** [comment]
Thanks for the review, I addressed the comments here: https://github.com/mosaicml/composer/pull/1491/commits/492cef36febab6a657b42a66d004bde41cc1d10d

**Review by mvpatel2000 — APPROVED** [review]
LGTM. Left one minor comment. I'll let you track and address the deepspeed issue separately
